### PR TITLE
CassandraToBigtable classic template feature: Cassandra Writetime replication 

### PR DESCRIPTION
**Pull request by georgecma**
Add Cassandra writetime replication feature to CassandraToBigtable classic template. 

Cassandra writetime can now be replicated to Bigtable as Bigtable cell timestamp. For this feature to work, users have to upload a json-like schema file to a GCS bucket location first for the table they wish to replicate writetimes for. The command to generate this schema file is:

```
cqlsh -e "select json * from system_schema.columns
                where keyspace_name='$CASSANDRA_KEYSPACE'
                and table_name='$CASSANDRA_TABLE'"
    > column_schema.json
```
Then, to upload the file:
```
gcloud storage cp column_schema $WRITETIME_CASSANDRA_COLUMN_SCHEMA
```
The schema GCS file path should then be set as $WRITETIME_CASSANDRA_COLUMN_SCHEMA to be parsed by the template at run time. 
 
The template replication behavior now sets Bigtable cell time as replication time (i.e. now) instead of epoch start time previously. SetZeroTimestamp is the backwards compatibility option to set 0 as the timestamp. 

@GoogleCloudPlatform/dataflow-templates-wg

### CODE REVIEW

File: v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraToBigtable.java
Comment (georgecma) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraToBigtable.java:243]: CassandraWritetimeQueryProvider is not getting unpacked in directRunner.

Comment (georgecma) [reply v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraToBigtable.java:243]: There's a bug in Beam cassandraIO: https://github.com/apache/beam/issues/31622

I'm working on a fix.

Comment (georgecma) [reply v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraToBigtable.java:243]: https://github.com/apache/beam/pull/31623 fix waiting for review and Beam release. Integration testing can continue after this beam release.

**Comment by georgecma** [comment]
@GoogleCloudPlatform/dataflow-templates-wg

File: v1/src/main/java/com/google/cloud/teleport/util/GCSAwareValueProvider.java
Comment (Polber) [inline v1/src/main/java/com/google/cloud/teleport/util/GCSAwareValueProvider.java:68]: This error message was intended to print failed path upon errors, but with this change, it will likely print null or empty string rather than the original path value.

This should be refactored to store `originalValue`

File: it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/bigtable/matchers/BigtableAsserts.java
Comment (Polber) [inline it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/bigtable/matchers/BigtableAsserts.java:94]: This class is intended for truthmatchers for making resource-dependent (in this case Bigtable) asserts.

Let's move these helper methods to the IT file they are being used in, or at the very least, create a utils class in the module that these methods are intended for (`v1/src/test/bigtable/...`)

**Review by Polber — CHANGES_REQUESTED** [review]
Thanks for the contribution! I left a couple small comments for core templates related files. Please tag a member from the BigTable team to review BigTable-related changes

File: v1/src/main/java/com/google/cloud/teleport/util/GCSAwareValueProvider.java
Comment (georgecma) [reply v1/src/main/java/com/google/cloud/teleport/util/GCSAwareValueProvider.java:68]: I refactored originalValue to returnValue to better reflect the variable potentially being changed in resolve(). 
This error will print out the GCS path - happy to change this variable name if there's some better name!

File: it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/bigtable/matchers/BigtableAsserts.java
Comment (georgecma) [reply it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/bigtable/matchers/BigtableAsserts.java:94]: Moved functions to CassandraToBigtableIT since that's the only place that they are used in.

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java
Comment (vermas2012) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java:252]: we are silently ignoring the timestamp column here, should we be raising the flag here? There are 2 cases where this is not good:

1. User's column names have special write time token which should be avoided, how do we inform them that its not a good practice? 
2. There is some bug in the regex or column name has special character that we are not handling well. We should raise a warning here, instead of silently ignoring the timestmap.

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java
Comment (vermas2012) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java:277]: This is a flawed assumption, we can't be sure of this. If I have a cell timestamp in in millis for future (lets say a day), you assume its in micro seconds and trim its last 3 digits. we lose last 3 digits of the millis and format is still wrong.

File: v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java
Comment (vermas2012) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java:68]: why? does cassandra not guarantee the output file format?

File: v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java
Comment (vermas2012) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java:99]: so regular column names have negative position values, and only the primary key have non negative position values?

File: v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java
Comment (vermas2012) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java:1]: I am very concerned about us using the regex for this parsing, if CQL evolves the output of the schema command, how do we adapt to it before a customer complains? Can you check if there is a cassandra library to parse the schema?

File: v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraToBigtable.java
Comment (vermas2012) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraToBigtable.java:202]: add the else: defaults to writetime=now.

File: v1/src/test/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFnTest.java
Comment (vermas2012) [inline v1/src/test/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFnTest.java:824]: Add a test with 0 timestamp
Add a test with timestamp = System.currentTimeInMillis() + 1000 and see what the timestamp comes out in CBT, is it same?

**Review by vermas2012 — CHANGES_REQUESTED** [review]

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java
Comment (georgecma) [reply v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java:252]: Added a warning

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java
Comment (georgecma) [reply v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java:277]: Added a 100 years of buffer tiem to parse millisecond timestamps.

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java
Comment (georgecma) [reply v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java:285]: The case here now is that this time is > 100 years into the future in microsecond timestamps. At this point, I think it'll be more important to call out this behavior than to modify it since we're not sure what the intention is anymore.

File: v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java
Comment (georgecma) [reply v1/src/main/java/com/google/cloud/teleport/bigtable/CassandraColumnSchema.java:99]: Correct - documentation on this is non-existent but the code points to this: https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/schema/ColumnMetadata.java#L92
This has been the case at least for the last 9 years (at least by blame history)

File: v1/src/test/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFnTest.java
Comment (georgecma) [reply v1/src/test/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFnTest.java:824]: Added tests for these scenarios.

File: v1/README_Cassandra_To_Cloud_Bigtable.md
Comment (vermas2012) [inline v1/README_Cassandra_To_Cloud_Bigtable.md:39]: > cqlsh -e "select json * from system_schema.columns where keyspace_name='$CASSANDRA_KEYSPACE' and table_name='$CASSANDRA_TABLE'

What happens if you execute this query from the cassandra client and interpret the results from resultset? I find it hard to behelive that we can query using cqlsh but can't fetch this data via an API. 

I have suffered a lot from maintaining compatiblity of hbase clients. I don't want to add similar pain by parsing strings in cassandra. Lets see if there is any way to do this via API contract

Comment (georgecma) [reply v1/README_Cassandra_To_Cloud_Bigtable.md:39]: Discussed offline - will keep the current approach. 

The result of the cqlsh command will be like [the following](https://docs.datastax.com/en/cql-oss/3.x/cql/cql_using/useQueryJSON.html): 
```
cqlsh:cycling> select json name, checkin_id, timestamp from checkin;
 [json]
------------------------------------------------------------------------------------------------------------------
 {"name": "BRAND", "checkin_id": "50554d6e-29bb-11e5-b345-feff8194dc9f", "timestamp": "2016-08-28 21:45:10.406Z"}
  {"name": "VOSS", "checkin_id": "50554d6e-29bb-11e5-b345-feff819cdc9f", "timestamp": "2016-08-28 21:44:04.113Z"}
(2 rows)
```

Note the prefix `[json] ---...` and suffix `(2 rows)` strings that will interfere with strict parsing. I wasn't able to find a library that does fuzzy JSON matching, so we have our own JSON parser for this.

Alternatively, datastax does have a systems querier to get system column table info. The ideal scenario would be to slot a datastax connector as the first step in the pipeline to process Cassandra table schema, and then feed the results to CassandraIO. However:
1. CassandraIO [does not allow any processors before itself in a pipeline](https://beam.apache.org/releases/javadoc/2.29.0/org/apache/beam/sdk/io/cassandra/CassandraIO.Read.html)
2. It is not possible to preprocess steps and dynamically construct a pipeline in a V1 template

In a future iteration, we should either loosen CassandraIO.Read or upgrade this template to a V2 template so we can have the writetimeCassandraColumnSchema automatically processed. Added a TODO for myself to explore these approached at a future time.

**Review by vermas2012 — APPROVED** [review]
Please go through all the warnings of lines not covered by tests and add any relevant Tests.

**Review by Abacn — COMMENTED** [review]
ack. will merge after fixing the format (mv. spotless:apply)

**Review by vermas2012 — APPROVED** [review]

**Comment by Abacn** [comment]
cc: @Polber looks like merge is still blocked pending your approval

**Comment by vermas2012** [comment]
@Polber Is there anything pending from the CBT side? can we merge this change?

**Review by Polber — APPROVED** [review]
Thanks!

**Pull request merged**
